### ISSUE TITLE
Training time for CIFAR-10 is higher on multi GPU cards compared to single

### ISSUE THREAD

**Issue by dashrathc** [tensorflow/tensorflow#4272]
When running CIFAR-10 example on two GPU cards in a single machine, the training time is higher compared to a single GPU card.

It is utilizing both cards but overall execution time is higher compared to single GPU.

![multi-gpu-cards](https://cloud.githubusercontent.com/assets/7776101/18342617/48dfeb66-75cd-11e6-83cb-473be9746f3a.png)

One more thing I noticed is that the queue is filled twice. I got following messages in log.
`Filling queue with 20000 CIFAR images before starting to train. This will take a few minutes.`
`Filling queue with 20000 CIFAR images before starting to train. This will take a few minutes.`

Ideally, It should be filled once and distribute across two GPUs but that is not happening. 
Could anyone please help me out to resolve this?

**Comment by tatatodd** [comment]
@dashrathc Please fill out the full information from the template that appears when you click on the green "New issue" button, including OS, libcuda info, versions, etc.  Thanks!

**Comment by dashrathc** [comment]
### Environment info
Operating System: Ubuntu 16.04.1 LTS

Installed version of CUDA and cuDNN: 
![libcudnn](https://cloud.githubusercontent.com/assets/7776101/18408249/11279cf6-7748-11e6-9044-3c7eb081b7fb.png)

TensorFlow version:
![tensorflow_version](https://cloud.githubusercontent.com/assets/7776101/18408261/51d09c12-7748-11e6-9f74-bc4a24003c93.png)

git rev-parse HEAD
8a3107801d15bf8af36221ff5bca0b94bf44d6d3

bazel version:
Build target: bazel-out/local-fastbuild/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Thu Jan 01 00:00:00 1970 (0)
Build timestamp: Thu Jan 01 00:00:00 1970 (0)
Build timestamp as int: 0

@tatatodd  let me know if need anything else.

**Comment by JianbangZ** [comment]
@dashrathc 
Hi, having two queues is fine. It may consume CPU a little bit more, but from my experience, at least for two GPU cards, it works equivalent as one queue.
One thing to remember is too adjust the batch_size correctly. For the cifar10 example, batch_size is the size for each GPU card not the overall batch size. So if you are using batch_size=128 for one GPU, you should change batch_size=64 for two GPUs to notice the speed difference.

**Comment by dashrathc** [comment]
@JianbangZ Thanks for reply.
I have updated 128 to 64 cifar10.py file.
tf.app.flags.DEFINE_integer('batch_size', 128, """Number of images to process in a batch.""")

Is it how batch_size should be changed? Please do let me know if there is any other way to update it.

**Comment by JianbangZ** [comment]
@dashrathc It is correct

**Comment by dashrathc** [comment]
@JianbangZ 
I have run it again for 100k iterations and following is the result,
batch_size=128, Elapsed Time: 18188.31 seconds
batch_size=64, Elapsed Time: 18072.72 seconds

The elapsed time difference is very minimal though batch_size decreased to by 64.
Could you please share your view?

**Comment by asimshankar** [comment]
@dashrathc : Are you still looking into this?

You might find this comment: https://github.com/tensorflow/tensorflow/issues/4744#issuecomment-252088066 interesting. There are many factors to consider when using multiple GPUs.

Have you seen this tutorial: https://www.tensorflow.org/tutorials/deep_cnn/index.html?
In particular, it talks about both single GPU and multi GPU training. The training is slightly different between the two as seen in the files [cifar10_train.py](https://github.com/tensorflow/tensorflow/blob/r0.11/tensorflow/models/image/cifar10/cifar10_train.py) and [cifar10_multi_gpu_train.py](https://github.com/tensorflow/tensorflow/blob/r0.11/tensorflow/models/image/cifar10/cifar10_multi_gpu_train.py). In particular, see the comment at the top of  [cifar10_multi_gpu_train.py](https://github.com/tensorflow/tensorflow/blob/r0.11/tensorflow/models/image/cifar10/cifar10_multi_gpu_train.py) which outlines the observed training time when using 1 GPU and 2 GPUs.

Let us know if this was helpful or not.

**Comment by asimshankar** [comment]
Closing due to inactivity, feel free to reopen if your concerns haven't been addressed and the pointers above didn't help.

**Comment by abrahamrhoffman** [comment]
Thanks! This was helpful :)
